sourcegraph/sourcegraph
# [App Setup]: Add new local repositories flow

Part of https://github.com/sourcegraph/sourcegraph/issues/51251

## Background

This PR adds a new UI for folders and repositories list to the new setup setting page; with the new UI, you can
- See folder structure by a given path
- Delete folder/repositories (note that you can't delete specific repositories from a folder, you have to delete the whole directory, it's a technical limitation of the current API) 

| State | Light | Dark | 
| ----- | ------ | ------ | 
| Zero state | <img width="1049" alt="Screenshot 2023-05-10 at 18 59 06" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/17c9c15d-b620-4e16-8df1-1c0cd37d098b"> | <img width="1136" alt="Screenshot 2023-05-10 at 19 30 33" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/1e9e0f2f-4d97-4b99-a472-d922a6adb84b"> | 
| With added repositories | <img width="1049" alt="Screenshot 2023-05-10 at 18 59 27" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/c7750b28-2cc4-4fd2-9a8d-3f6b20ab438b"> | <img width="1049" alt="Screenshot 2023-05-10 at 18 59 32" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/1a3b9b0d-0a9e-4741-87cf-17e49d236c84"> |

## Todos
- [x] Extract the common API part from the current local setup step 
- [x] Utilize existing API in reusable parts
- [x] Implement folder repositories logic on the client 
- [x] Polish new local repositories UI
- [x] Solve the problem with Multiple request race condition 

## Tech note
I'm not happy with our current API on the backend. I think to make client logic more resilient. We have to change our API. I described my suggestions in this doc. If you're changing something on the backed, I would suggest this doc first. https://docs.google.com/document/d/1Rn-9da0rELSnmQCVdO6rn_G9zzg-GH3PFFYIdZru6NE/edit#


## Test plan
- Check the UI functional (add repositories, directories, remove repositories, dirs)
- Check that this PR doesn't break nothing in the old Setup wizard local repositories step
